src/Test/TestCases.Runtime/MissingRequiredExtensionTests.cs

```diff
@@ -0,0 +1,60 @@
+﻿// This file is part of Core WF which is licensed under the MIT license.
+// See LICENSE file in the project root for full license information.
+
+using Shouldly;
+using System.Activities;
+using System.Activities.Statements;
+using System.Text;
+using Xunit;
+
+namespace TestCases.Runtime
+{
+    public class MissingRequiredExtensionTests
+    {
+        [Theory]
+        [InlineData(true)]
+        [InlineData(false)]
+        public void RunActivityWithRequiredExtension(bool registerExtension)
+        {
+            var sequence = new Sequence
+            {
+                Activities = { new ActivityWithRequiredExtension(registerExtension) }
+            };
+
+            var exception = Record.Exception(new WorkflowApplication(sequence).Run);
+
+            if (registerExtension)
+            {
+                exception.ShouldBeNull();
+            }
+            else
+            {
+                exception.ShouldBeOfType<ExtensionRequiredException>();
+                ((ExtensionRequiredException)exception).RequiredExtensionType.ShouldBe(typeof(StringBuilder));
+            }
+        }
+
+        private class ActivityWithRequiredExtension : NativeActivity
+        {
+            private readonly bool _registerExtension;
+
+            public ActivityWithRequiredExtension(bool registerExtension)
+            {
+                _registerExtension = registerExtension;
+            }
+
+            protected override void CacheMetadata(NativeActivityMetadata metadata)
+            {
+                metadata.RequireExtension<StringBuilder>();
+                if (_registerExtension)
+                {
+                    metadata.AddDefaultExtensionProvider(() => new StringBuilder());
+                }
+            }
+
+            protected override void Execute(NativeActivityContext context)
+            {
+            }
+        }
```

Comment:
```suggestion
private class ActivityWithRequiredExtension<T> : NativeActivity where T : class, new()
{
    private readonly bool _registerExtension;

    public ActivityWithRequiredExtension(bool registerExtension)
    {
        _registerExtension = registerExtension;
    }

    protected override void CacheMetadata(NativeActivityMetadata metadata)
    {
        metadata.RequireExtension(typeof(T));
        if (_registerExtension)
        {
            metadata.AddDefaultExtensionProvider(() => new T());
        }
    }

    protected override void Execute(NativeActivityContext context)
    {
    }
}
```
`ActivityWithRequiredExtension` can be kep generic if `T` has constrains to be a reference type and to have a parameterless constructor.

Reply:
Line `21` will become:
`Activities = { new ActivityWithRequiredExtension<StringBuilder>(registerExtension) }`

Reply:
But we only use it in this single test and we're not interested in testing different extension types, so I'd keep it simple and hardcode a type.